### ISSUE TITLE
Add torch.export.ExportedProgram (.pt2) as an export format

### ISSUE THREAD

**Issue by AyoubMDL** [ultralytics/ultralytics#24124]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hello,

I want to know if it is possible to add support for exporting models to `torch.export.ExportedProgram` (.pt2) format (like model.export(format="exported_program")). Ultralytics already produces an ExportedProgram internally as part of the ExecuTorch export path:
```python
# ultralytics/utils/export/executorch.py
et_program = to_edge_transform_and_lower(
    torch.export.export(model, (sample_input,)),  #  ExportedProgram created here
    partitioner=[XnnpackPartitioner()],
).to_executorch()
```
However, there is no way to obtain the ExportedProgram on its own.

### Use case

`torch.export.ExportedProgram` is the standard entry point for downstream compilers and runtimes in the PyTorch ecosystem. It is used to apply graph transforms and to lower to formats like MLIR...

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @AyoubMDL, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your feature request triaged quickly, and an Ultralytics engineer will also assist soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.
Since this is a feature request ✨, it would be very helpful if you could share a bit more implementation detail, such as your target workflow, expected API behavior, and any notes on how you’d like `.pt2` / `torch.export.ExportedProgram` support to fit into the existing export interface.
If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
@onuralpszr

**Comment by reallyticsai** [comment]
Sure, it's definitely possible to add support for exporting models in the `torch.export.ExportedProgram` format. Given our experience with deploying models in production, we've found that this format is particularly useful for downstream compilers and runtimes.

Here's a concrete suggestion:

1. **Modify the Export Function**: Update the export function to include an option for exporting models as `torch.export.ExportedProgram`. This can be done by adding a new parameter to the existing `model.export` function.

2. **Handle the ExportedProgram**: After creating the `ExportedProgram`, you can then use it for further transformations or lowering to formats like MLIR.

Here's a rough outline of what the code might look like:

```python
def export_model(model, input_sample, format='default'):
    if format == 'exported_program':
        et_program = to_edge_transform_and_lower(
            torch.export.export(model, (input_sample,)),
            partitioner=[XnnpackPartitioner()],
        ).to_executorch()
        return et_program
    # Add other export formats here
```

This would allow users to specify the format they want to export to, including `exported_program`.

If you'd like to dive deeper into the specifics, let me know!

**Comment by glenn-jocher** [comment]
This looks reasonable to add; the current [ExecuTorch export flow](https://docs.ultralytics.com/reference/engine/exporter/#ultralytics.engine.exporter.Exporter.export_executorch) creates an `ExportedProgram` internally and immediately lowers it to `.pte`, so supporting `.pt2` cleanly would need a separate export path that saves the raw `torch.export.ExportedProgram` before lowering. If you’d like to open a PR, please keep it minimal and scoped to this format, avoid adding a new `default.yaml` arg, and include docs plus a small export test.

**Comment by Harshad071** [comment]
Hi @glenn-jocher, I’d like to work on this.

Based on your suggestion to keep it minimal, I’m planning to implement a separate export path that:

- Adds `format="pt2"` (aligned with `.pt2` extension)
- Uses `torch.export.export()` to generate the ExportedProgram
- Saves the raw ExportedProgram using `torch.export.save()`
- Avoids any ExecuTorch lowering or backend changes
- Keeps the scope limited to export + a small test

This would provide a clean intermediate representation for downstream workflows without modifying existing export pipelines.

Let me know if this approach looks good , I can open a PR shortly.

**Comment by glenn-jocher** [comment]
Yes, that approach looks good—please open a minimal PR with `format="pt2"` that saves a raw `torch.export.ExportedProgram` via `torch.export.save()` to `file.with_suffix(".pt2")`, following the single-artifact pattern in the [TorchScript export reference](https://docs.ultralytics.com/reference/utils/export/torchscript/) rather than the lowered [ExecuTorch path](https://docs.ultralytics.com/reference/utils/export/executorch/), and include a small export test plus docs.
